File: CPP/hashmap/17. Letter Combinations of a Phone Number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        unordered_map<char, vector<string>> map;
        map['2'] = {"a","b","c"};
        map['3'] = {"d","e","f"};
        map['4'] = {"g","h","i"};
        map['5'] = {"j","k","l"};
        map['6'] = {"m","n","o"};
        map['7'] = {"p","q","r","s"};
        map['8'] = {"t","u","v"};
        map['9'] = {"w","x","y","z"};

        vector<string> temp;
        vector<string> temp_res;
        vector<string> res;
        for (int i = 0; i < digits.size(); i++){
            for (string c: map[digits[i]]){
                temp.push_back(c);
            }
            if (res.size() == 0) res = temp;
            else {
                for (string sr: res){
                    for (string st: temp) {
                        temp_res.push_back(sr + st);
                    }                    
                }
                res = temp_res;
                temp_res.clear();
            }
            temp.clear();
        }
        return res;
    }
};

File: Python/Hashmap/17. Letter Combinations of a Phone Number.py
from typing import List
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        map = {"2":"abc", "3": "def","4": "ghi", "5": "jkl", "6":"mno", "7":"pqrs", "8": "tuv", "9": "wxyz"}
        stack = []
        res = []
        for n in digits:
            if n in map:
                if len(res) == 0:
                    for k in range(len(map[n])):
                        stack.append(map[n][k])
                else:
                    for m in res:
                        for k in range(len(map[n])):
                            stack.append(m + map[n][k])
                res = stack
                stack = []
                        
        return res

if __name__ == '__main__':
    digits = "23"
    sol = Solution()
    print(sol.letterCombinations(digits))